### PR TITLE
Detect and output a better hint for cyclic dependencies

### LINKED ISSUE #9837: DX: Separate error message for cyclic dependencies
I get asked from time to time from less experienced composer users why they have errors like

grafik

the actual problem is, that they have defined a cyclic dependency without realizing it.
would it make sense/feasable to use/print a error/warning message which explicitly states that?

cyclic dependencies lead to a lot of problems, and leaving less experienced people with a error message which they need to interpret and realize that they actually build a circle is not helpful.

maybe its not possible or hard to identify - I have no idea.. just leaving the idea here, to gather some feedback.

### PR DESCRIPTION
**Pull request by Seldaek**
Fixes #9837

### CODE REVIEW

File: doc/articles/troubleshooting.md [hunk 1]
```diff
@@ -47,32 +47,45 @@ This is a list of common pitfalls on using Composer, and how to avoid them.
    In this case add the `--with-dependencies` argument **or** add all dependencies which
    need an update to the command.
 
-## Package not found on travis-ci.org
+## Cyclic / circular dependencies
```
Comment (naderman) [inline doc/articles/troubleshooting.md:50]: Cycling or circular dependencies are a normal thing and don't cause problems. What this describes is rather a Cyclic / circular dependency including the root package, so I think the title too should make it clear this is only about situations involving the root package.

File: doc/articles/troubleshooting.md [hunk 2]
```diff
@@ -47,32 +47,45 @@ This is a list of common pitfalls on using Composer, and how to avoid them.
...
 
-1. Check the ["Package not found"](#package-not-found) item above.
+Cyclic (or circular) dependency occurs when your root package depends on a package
+which ends up depending (directly or indirectly) back on the root package itself.
+
+The best is to avoid this really if possible as it can lead to problems, but if you
+absolutely must have a cyclic dependency read on to figure out solutions.
+
+Cyclic dependencies commonly causes issues in two cases:
```
Comment (naderman) [inline doc/articles/troubleshooting.md:58]: Here to you need to be clear, cyclic dependencies do not typically cause any of these problems, a cyclic dependency on the root package may.

File: src/Composer/DependencyResolver/Problem.php [hunk 3]
```diff
@@ -420,6 +421,16 @@ private static function computeCheckForLowerPrioRepo($isVerbose, $packageName, $
             }
         }
 
+        if ($higherRepoPackages) {
+            $topPackage = reset($higherRepoPackages);
+            if ($topPackage instanceof RootPackageInterface) {
+                return array(
+                    "- Root composer.json requires $packageName".self::constraintToText($constraint).', it is ',
+                    'satisfiable by '.self::getPackageList($nextRepoPackages, $isVerbose).' from '.$nextRepo->getRepoName().' but the root package can not be updated. This is due to a cyclic dependency. See https://getcomposer.org/cyclic-dep for details and assistance.'
```
Comment (naderman) [inline src/Composer/DependencyResolver/Problem.php:429]: This is confusing, what is due to a cyclic dependency? The root package cannot be updated?

**Review by naderman — CHANGES_REQUESTED** [review]
I actually think this should entirely be rewritten to put less of a focus on cyclic/circular dependencies and instead make it clear that a dependency on something which depends on a specific root package version is the problem.

File: tests/Composer/Test/Fixtures/installer/circular-dependency-errors.test [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+2
+
+--EXPECT-OUTPUT--
+Loading composer repositories with package information
+Updating dependencies
+Your requirements could not be resolved to an installable set of packages.
+
+  Problem 1
+    - Root composer.json requires requires/root 1.0.0 -> satisfiable by requires/root[1.0.0].
+    - requires/root 1.0.0 requires root/pkg ^1.0 -> satisfiable by root/pkg[1.0.0] from package repo (defining 2 packages) but the root package can not be updated so it is not installable. See https://getcomposer.org/cyclic-dep for details and assistance.
```
Comment (naderman) [inline tests/Composer/Test/Fixtures/installer/circular-dependency-errors.test:43]:
```suggestion
    - requires/root 1.0.0 requires root/pkg ^1.0 -> satisfiable by root/pkg[1.0.0] from package repo (defining 2 packages) but root/pkg is the root package and cannot be modified. See https://getcomposer.org/cyclic-dep for details and assistance.
```

**Review by naderman — APPROVED** [review]

**Comment by Seldaek** [comment]
Added https://github.com/composer/composer/commit/0dce0f80f12543304fb0f1f755fb7bf9dbecfc9e to this :)

**Pull request merged**